### DISCUSSION TITLE
`TypeIs` narrows constrained type variable to `Never`

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#7759] [General]
Minimal reproducible example ([playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigBVEsBJQgKHYBMthzCNCYIABQA3AFxQwAIwBWWAMYwAlFAC0APnqMWAbQEgAumNZRTUAHSX2rLj2BgwOuhKH6ANORTKDoiXRUa9MZm5DxIfPqiSsEhZiBYIlgAhgA2GPAIWFGmAMRQAHIJWCAmsabxMACuIChQIqWmWCmEWDFl8Ymp6YzZUHl0AAL2YA2xFdW19UA)):

```python
from typing_extensions import TypeIs

def is_str(v: object) -> TypeIs[str]:
    ...

def foo[T: (str, int)](v: T) -> T:
    if is_str(v):
        reveal_type(v)  # Never
		return v
    ...
```

Is $(str, int) \land str$ really $\emptyset$?

**Comment by erictraut** [comment]
I agree this is a bug. Converting to a bug report so it's tracked accordingly.

**Reply by erictraut** [reply]
Here's a link to the issue: https://github.com/microsoft/pyright/issues/7760

**Reply by minmax** [reply]
IMHO version 1.1.361 does not match the description of [PEP 742](https://peps.python.org/pep-0742/) for `else` branches.

Is it possible to add support for these in `isinstance`, so that the new version of `TypeIs` will also support them?

From PEP: "TypeIs can narrow the type in both the if and else branches of a conditional."

"When a TypeIs function returns False, type checkers can narrow the type of the variable to exclude the TypeIs type."

**Reply by erictraut** [reply]
Please file a new issue.

**Reply by minmax** [reply]
Here's a link to the issue: https://github.com/microsoft/pyright/issues/7837
